File: packages/rails/lib/korea_auth/providers/google.rb
# frozen_string_literal: true

require "httparty"
require "jwt"

module KoreaAuth
  module Providers
    class Google < Base
      TOKEN_INFO_URL = "https://oauth2.googleapis.com/tokeninfo"

      def verify(token)
        # For ID tokens, we can verify with Google's tokeninfo endpoint
        response = HTTParty.get("#{TOKEN_INFO_URL}?id_token=#{token}")
        
        unless response.success?
          return Result.failure("Invalid token")
        end

        data = response.parsed_response

        Result.success(
          uid: data["sub"],
          provider: "google",
          name: data["name"],
          email: data["email"],
          photo_url: data["picture"],
          raw_data: data
        )
      rescue StandardError => e
        Result.failure(e.message)
      end
    end
  end
end

File: packages/flask/korea_auth/providers/google.py
"""
Google provider implementation
"""

import requests
from ..result import AuthResult


class GoogleProvider:
    """Verify Google login tokens."""
    
    TOKEN_INFO_URL = "https://oauth2.googleapis.com/tokeninfo"
    
    def verify(self, token: str) -> AuthResult:
        """Verify a Google ID token and return user info."""
        try:
            response = requests.get(f"{self.TOKEN_INFO_URL}?id_token={token}")
            
            if not response.ok:
                return AuthResult.failure("Invalid token")
            
            data = response.json()
            
            return AuthResult.success_result(
                uid=data["sub"],
                provider="google",
                name=data.get("name"),
                email=data.get("email"),
                photo_url=data.get("picture"),
                raw_data=data,
            )
        except Exception as e:
            return AuthResult.failure(str(e))
